wagtail/core/models.py

```diff
@@ -36,9 +39,117 @@
+            return
+        return cache
+
+    @staticmethod
+    def get_hostname_and_port_from_request(request):
+        """Reliably extracts 'hostname' and 'port' values from the supplied
+        ``HttpRequest`` instance (without raising exceptions), and return them
+        as a tuple"""
+        try:
+            hostname = request.get_host().split(':')[0]
+        except KeyError:
+            hostname = None
+        try:
+            port = request.get_port()
+        except KeyError:
+            port = None
+        return hostname, port
+
+    def _get_keys_for_unique_hostname(self, hostname):
+        """Returns a tuple of additional keys to use when populating the cache
+        for a ``Site`` with a unique hostname. These keys allow requests to be
+        matched to the site even if the request's port value differs from the
+        site's saved ``port`` field value"""
+        valid_ports = getattr(settings, 'WAGTAIL_SITE_CACHE_VALID_PORTS',
+                              (80, 443, 8000, 8080, None))
+        return tuple("{0}:{1}".format(hostname, port) for port in valid_ports)
+
+    def clear_cache(self):
+        cache = self.get_site_cache()
+        if cache is not None:
+            cache.delete(SITE_CACHE_SITES_KEY)
+
+    def populate_cache(self):
+        """Populates the cache with a dictionary of ``Site`` objects, keyed
+        with hostname/port combinations, which get_for_request() will use
+        to match ``HttpRequest`` objects to the relevant ``Site``"""
+        cache = self.get_site_cache()
+        if cache is None:
+            return
+
+        sites = {}
+        unique_hostname_sites = {}
+        for site in self.all().select_related('root_page').iterator():
+
+            sites["{0}:{1}".format(site.hostname, site.port)] = site
+            if site.is_default_site:
+                sites[SITE_CACHE_DEFAULT_KEY] = site
+
+            if site.hostname in unique_hostname_sites:
+                del unique_hostname_sites[site.hostname]
+            else:
+                unique_hostname_sites[site.hostname] = site
+
+        for hostname, site in unique_hostname_sites.items():
+            additional_keys = self._get_keys_for_unique_hostname(hostname)
+            sites.update({key: site for key in additional_keys})
+
+        cache.set(SITE_CACHE_SITES_KEY, sites, None)
+        return sites
+
+    def get_for_request(self, request):
+        """Return the site responsible for dealing with the supplied
+        ``HttpRequest`` instance"""
+        hostname, port = self.get_hostname_and_port_from_request(request)
+        cache = self.get_site_cache()
+        if cache is None:
+            return get_site_for_hostname(hostname, port)
+
+        sites = cache.get(SITE_CACHE_SITES_KEY) or self.populate_cache()
+        request_key = "{0}:{1}".format(hostname, port)
+        return sites.get(request_key) or sites.get(SITE_CACHE_DEFAULT_KEY)
```

Comment:
Does falling back on the default site like this subtly change the existing behavior in the case where there are multiple sites but the incoming request doesn't match any of them? [In `get_site_for_hostname` here](https://github.com/wagtail/wagtail/blob/master/wagtail/core/sites.py#L49), the logic seems to raise `Site.DoesNotExist` if there are multiple sites but the request doesn't match any. With this change, the logic would return the default site instead. Am I misreading?

Reply:
@chosak Thanks. Yes, I think you could be right here. I will review to make sure this is fully backwards compatible.